{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Install Libraries\n",
    "Install information can be found [here](https://www.archrproject.com/index.html)\\\n",
    "Vignette [here](https://www.archrproject.com/articles/Articles/tutorial.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "#ArchR parameters\n",
    "atac_frag = \"\" # Input file  \n",
    "genome = \"\" #either hg38 or mm10\n",
    "peak_set = \"\" #bed file\n",
    "\n",
    "#ArchR QC\n",
    "min_tss = 4 #The minimum numeric transcription start site (TSS) enrichment score required for a cell to pass filtering\n",
    "min_frags = 1000 #The minimum number of mapped ATAC-seq fragments required per cell to pass filtering for use\n",
    "add_tile_mat = TRUE #A boolean value indicating whether to add a \"Tile Matrix\" to each ArrowFile. \n",
    "add_gene_score_mat = TRUE #A boolean value indicating whether to add a Gene-Score Matrix to each ArrowFile.\n",
    "\n",
    "#ArchR Doublet paramaters\n",
    "find_doublets = FALSE\n",
    "doublet_k = 10 #The number of cells neighboring a simulated doublet to be considered as putative doublets.\n",
    "doublet_knn_method = \"UMAP\" #Refers to the embedding to use for nearest neighbor search.\n",
    "lsi_method = 1 #A number or string indicating the order of operations in the TF-IDF normalization. Possible values are: 1 or \"tf-logidf\", 2 or \"log(tf-idf)\", and 3 or \"logtf-logidf\". \n",
    "\n",
    "copy_arrow_files = TRUE #save a copy of arrow files in the ArchR project (recommended)\n",
    "iter_LSI_matrix = \"TileMatrix\" #The name of the data matrix to retrieve from the ArrowFiles associated with the ArchRProject. Valid options are \"TileMatrix\" or \"PeakMatrix\".\n",
    "threads = 1\n",
    "prefix = \"prefix\" #project name\n",
    "\n",
    "#ArchR Plots parameters\n",
    "marker_features_test = \"wilcoxon\" #The name of the pairwise test method to use in comparing cell groupings to the null cell grouping during marker feature identification.\n",
    "heatmap_transpose = TRUE #Boolean to transpose heatmap\n",
    "heatmap_label_n = 5 #Top n genes to label per cluster in heatmap\n",
    "heatmap_cutoff = \"FDR <= 0.01 & Log2FC >= 0.5\" #Cut-off applied to genes in heatmap\n",
    "\n",
    "#Terra specific parameters\n",
    "table_name = \"demux_BH3KTLDMXY\"\n",
    "experiment_name = \"gm12878_fresh_ATAC\"\n",
    "\n",
    "#papermill specific parameters\n",
    "papermill = TRUE\n",
    "\n",
    "#jupyter notebook plot sizes\n",
    "options(repr.plot.width=20, repr.plot.height=15)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "papermill <- as.logical(papermill)\n",
    "add_tile_mat <- as.logical(add_tile_mat)\n",
    "add_gene_score_mat <- as.logical(add_gene_score_mat)\n",
    "copy_arrow_files <- as.logical(copy_arrow_files)\n",
    "heatmap_transpose <- as.logical(heatmap_transpose)\n",
    "find_doublets <- as.logical(find_doublets)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "r-libraries"
    ]
   },
   "outputs": [],
   "source": [
    "if (!requireNamespace(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\", repos = \"https://cran.r-project.org\")\n",
    "if (!requireNamespace(\"devtools\", quietly = TRUE)) \n",
    "    install.packages(\"devtools\")\n",
    "if (!requireNamespace(\"GenomeInfoDbData\", quietly = TRUE))\n",
    "    BiocManager::install(\"GenomeInfoDbData\")\n",
    "if (!requireNamespace(\"GenomicRanges\", quietly = TRUE))\n",
    "    BiocManager::install(\"GenomicRanges\")\n",
    "if (!requireNamespace(\"Rsamtools\", quietly = TRUE))\n",
    "    BiocManager::install(\"Rsamtools\")\n",
    "if (!requireNamespace(\"presto\", quietly = TRUE))\n",
    "    devtools::install_github('immunogenomics/presto')\n",
    "if (!requireNamespace(\"magick\", quietly = TRUE))\n",
    "    install.packages(\"magick\")\n",
    "if (!requireNamespace(\"Seurat\", quietly = TRUE))\n",
    "    install.packages(\"Seurat\")\n",
    "if (!requireNamespace(\"ArchR\", quietly = TRUE)){\n",
    "    devtools::install_github('GreenleafLab/ArchR@v1.0.1', repos = BiocManager::repositories())\n",
    "    ArchR::installExtraPackages()\n",
    "}\n",
    "if (!requireNamespace(\"logr\", quietly = TRUE))\n",
    "    install.packages(\"logr\")\n",
    "if (!requireNamespace(\"grid\", quietly = TRUE))\n",
    "    install.packages(\"grid\")\n",
    "if (!requireNamespace(\"gridExtra\", quietly = TRUE))\n",
    "    install.packages(\"gridExtra\")\n",
    "if (!requireNamespace(\"ggplot2\", quietly = TRUE))\n",
    "    install.packages(\"ggplot2\")\n",
    "if (!requireNamespace(\"hexbin\", quietly = TRUE))\n",
    "    install.packages(\"hexbin\")\n",
    "if (!requireNamespace(\"ggpointdensity\", quietly = TRUE))\n",
    "    install.packages(\"ggpointdensity\")\n",
    "\n",
    "options(\"logr.notes\" = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "suppressMessages(library(parallel))\n",
    "suppressMessages(library(ArchR))\n",
    "suppressMessages(library(magick))\n",
    "suppressMessages(library(logr))\n",
    "suppressMessages(library(GenomicRanges))\n",
    "suppressMessages(library(grid))\n",
    "suppressMessages(library(gridExtra))\n",
    "suppressMessages(library(ggplot2))\n",
    "suppressMessages(library(ggpointdensity))\n",
    "\n",
    "set.seed(1)\n",
    "addArchRThreads(threads = threads)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Function to save plots\n",
    "plot_filename = paste0(prefix,\".atac.archr.plots.\",genome)\n",
    "dir.create(plot_filename, showWarnings=F)\n",
    "printPNG <- function(name, plotObject, papermill, wf=22, hf=11){\n",
    "    filename = paste0(plot_filename,\"/\",prefix,\".atac.archr.\",name,\".\",genome,\".png\")\n",
    "    if(papermill){\n",
    "        ggsave(plot = plotObject, filename = filename, width = wf, height = hf)\n",
    "    }\n",
    "}\n",
    "\n",
    "#Function to create plots\n",
    "create_plot = function(plot_list, title, subtitle, heights=unit(c(1,1,8), rep(\"in\",3)), width = 7){\n",
    "    g = c(list(title),list(subtitle),plot_list)\n",
    "    N = length(plot_list)\n",
    "    laym = rbind(rep(1,N),rep(2,N),(3:(N+2)))\n",
    "    widths = unit(rep(width, length(plot_list)), rep(\"in\",length(plot_list)))\n",
    "    obj = arrangeGrob(grobs=g,layout_matrix=laym, heights=heights, widths = widths)\n",
    "    return(obj)\n",
    "}\n",
    "\n",
    "#Function to calculate density for scatterplot coloring \n",
    "get_density <- function(x, y, ...) {\n",
    "  dens <- MASS::kde2d(x, y, ...)\n",
    "  ix <- findInterval(x, dens$x)\n",
    "  iy <- findInterval(y, dens$y)\n",
    "  ii <- cbind(ix, iy)\n",
    "  return(dens$z[ii])\n",
    "}\n",
    "\n",
    "#Function to write h5 file from dgCMatrix\n",
    "#Taken from https://rdrr.io/github/AllenInstitute/scrattch.io/src/R/write_10x.R\n",
    "write_dgCMatrix_h5 <- function(mat,\n",
    "                               cols_are = \"gene_names\",\n",
    "                               h5_target,\n",
    "                               ref_name = \"prefix\",\n",
    "                               gene_ids = NULL) {\n",
    "\n",
    "  #library(Matrix)\n",
    "\n",
    "  if(grepl(\"gene\",cols_are)) {\n",
    "    mat <- Matrix::t(mat)\n",
    "  }\n",
    "\n",
    "  # Create target file\n",
    "  rhdf5::h5createFile(h5_target)\n",
    "  # Create data group\n",
    "  rhdf5::h5createGroup(h5_target,\n",
    "                       ref_name)\n",
    "\n",
    "  # Store sample ids (barcodes) and gene names\n",
    "  rhdf5::h5write(colnames(mat),\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/barcodes\"))\n",
    "  rhdf5::h5write(rownames(mat),\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/gene_names\"))\n",
    "\n",
    "  if(is.null(gene_ids)) {\n",
    "    gene_ids <- rownames(mat)\n",
    "  }\n",
    "\n",
    "  rhdf5::h5write(gene_ids,\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/genes\"))\n",
    "\n",
    "  # Store dimensions as shape\n",
    "  rhdf5::h5write(dim(mat),\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/shape\"))\n",
    "\n",
    "  # Store values from mat@x as data\n",
    "  rhdf5::h5createDataset(h5_target,\n",
    "                         paste0(\"/\",ref_name,\"/data\"),\n",
    "                         dims = length(mat@x),\n",
    "                         storage.mode = \"integer\",\n",
    "                         chunk = 1000,\n",
    "                         level = 4)\n",
    "  rhdf5::h5write(mat@x,\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/data\"))\n",
    "\n",
    "  # Store row indices from mat@i as indices\n",
    "  rhdf5::h5createDataset(h5_target,\n",
    "                         paste0(\"/\",ref_name,\"/indices\"),\n",
    "                         dims = length(mat@i),\n",
    "                         storage.mode = \"integer\",\n",
    "                         chunk = 1000,\n",
    "                         level = 4)\n",
    "  rhdf5::h5write(mat@i,\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/indices\"))\n",
    "\n",
    "  # Store column pointers from mat@p as indptr\n",
    "  rhdf5::h5write(mat@p,\n",
    "                 h5_target,\n",
    "                 paste0(\"/\",ref_name,\"/indptr\"))\n",
    "\n",
    "}\n",
    "\n",
    "#Create log file\n",
    "logfile <- file.path(paste0(prefix,\".atac.archr.logfile.\",genome,\".txt\"))\n",
    "lf <- log_open(logfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Terra specific code block\n",
    "get_file <- function(path){\n",
    "    dest <- getwd()\n",
    "    gsutil_cp(path, dest)\n",
    "    name <- basename(path)\n",
    "    return(name)\n",
    "}\n",
    "\n",
    "if (!papermill){\n",
    "    table <- avtable(table_name)\n",
    "    atac_frag <- get_file(table$atac_fragment_file_raw[table[, sprintf('%s_id', table_name)] == experiment_name])\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "download-genome"
    ]
   },
   "outputs": [],
   "source": [
    "#Download genome\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Download genome\")\n",
    "    \n",
    "        #Code start to download genome\n",
    "    \n",
    "        addArchRGenome(genome)\n",
    "    \n",
    "        #Code end to download genome\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Download genome\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Download genome\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "create-arrow-file"
    ]
   },
   "outputs": [],
   "source": [
    "#Create Arrow files\n",
    "\n",
    "ArrowFiles = tryCatch({\n",
    "        log_print(\"# Create arrow files\")\n",
    "    \n",
    "        #Code start to create arrow files\n",
    "    \n",
    "        ArrowFiles <- createArrowFiles(\n",
    "          inputFiles = atac_frag,\n",
    "          sampleNames = prefix,\n",
    "          minTSS = 0, \n",
    "          minFrags = 0,\n",
    "          addTileMat = add_tile_mat,\n",
    "          addGeneScoreMat = add_gene_score_mat\n",
    "        )\n",
    "    \n",
    "        #Code end to create arrow files\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Create arrow files\")\n",
    "        return(ArrowFiles)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Create arrow files\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "create-archr-project"
    ]
   },
   "outputs": [],
   "source": [
    "#Create Archr project\n",
    "\n",
    "proj = tryCatch({\n",
    "        log_print(\"# Create Archr project\")\n",
    "    \n",
    "        #Code start to create Archr project\n",
    "    \n",
    "        proj <- ArchRProject(\n",
    "          ArrowFiles = ArrowFiles, \n",
    "          outputDirectory = prefix,\n",
    "          copyArrows = copy_arrow_files, #This is recommened so that you maintain an unaltered copy for later usage.\n",
    "          showLogo = FALSE\n",
    "        )\n",
    "    \n",
    "        #Code end to create Archr project\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Create Archr project\")\n",
    "        return(proj)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Create Archr project\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre-filtered TSS Enrichment vs Unique Fragments QC plot\n",
    "xlabel_min = 0\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Pre-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "\n",
    "        # Code start to create pre-filtered TSS Enrichment vs Unique Fragments QC plot    \n",
    "        df <- as.data.frame(getCellColData(proj, select = c(\"nFrags\", \"TSSEnrichment\")))\n",
    "        \n",
    "        df$density = get_density(df$nFrags, df$TSSEnrichment, n = 100)\n",
    "        xlabel_min = min(df$nFrags)\n",
    "        obj = ggplot(data=df) + \n",
    "                geom_point(aes(x = nFrags, y = TSSEnrichment, color = density)) +  \n",
    "                scale_x_continuous(trans = \"log10\", breaks = scales::trans_breaks(\"log10\", function(x) 10^x), labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "                expand_limits(x = xlabel_min, y = 0) + \n",
    "                geom_hline(yintercept = min_tss, lty = \"dashed\") + \n",
    "                geom_vline(xintercept = min_frags, lty = \"dashed\") +\n",
    "                scale_colour_gradientn(colors = paletteContinuous(set = \"sambaNight\")) +\n",
    "                xlab(label =\"Unique Fragments\") + \n",
    "                ylab(label = \"TSS Enrichment\") + \n",
    "                annotation_logticks(sides = \"b\") +\n",
    "                theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=8.5))                                                                 \n",
    "           \n",
    "        tg <- textGrob('Pre-filtered TSS Enrichment vs Unique Fragments QC plot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \"Total Barcodes: \", nCells(proj), \"\\n\", \n",
    "                        \"Median Frags: \", median(proj$nFrags), \"\\n\", \n",
    "                        \"Median TSS Enrichment: \", median(proj$TSSEnrichment), \"\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "        plot_list = list(obj)\n",
    "        obj = create_plot(plot_list, tg, sg,heights=unit(c(1.5,1.5,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "        printPNG(name = \"prefiltered_tss_by_uniq_frags\", plotObject = obj, papermill = papermill, wf = 11, hf = 12)\n",
    "            \n",
    "        # Code end to create pre-filtered TSS Enrichment vs Unique Fragments QC plot \n",
    "\n",
    "        log_print(\"SUCCESSFUL: Pre-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Pre-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Pre-filtered Fragment Size Distribution plot\n",
    "tryCatch({\n",
    "        log_print(\"# Pre-filtered Fragment Size Distribution plot\")\n",
    "\n",
    "        # Code start to create pre-filtered Fragment Size Distribution    \n",
    "        obj = plotFragmentSizes(ArchRProj = proj) +\n",
    "        theme_gray()\n",
    "            \n",
    "        tg <- textGrob('Pre-filtered Fragment Size Distribution', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \"Total Barcodes: \", nCells(proj), \"\\n\", \n",
    "                        \"Total nFrags: \", paste(format(round(sum(proj$nFrags) / 1e6, 1), trim = TRUE), \"M\"), \"\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "        plot_list = list(obj)\n",
    "        obj = create_plot(plot_list, tg, sg,heights=unit(c(1.5,1.5,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "        printPNG(name = \"prefiltered_frag_size_dist\", plotObject = obj, papermill = papermill, wf = 11, hf = 12)\n",
    "            \n",
    "        # Code end to create pre-filtered Fragment Size Distribution \n",
    "\n",
    "        log_print(\"SUCCESSFUL: Pre-filtered Fragment Size Distribution plot\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Pre-filtered Fragment Size Distribution plot\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Adding Peak set\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Adding Peak set\")\n",
    "    \n",
    "        #Code start to add peak set\n",
    "    \n",
    "        peakset = read.table(peak_set, header=F, sep = \"\\t\", stringsAsFactors = F)\n",
    "        colnames(peakset) = c(\"chr\",\"start\",\"end\")\n",
    "        gr_peakset = makeGRangesFromDataFrame(peakset)\n",
    "        gr_peakset = gr_peakset[seqnames(gr_peakset) %in% seqnames(getChromSizes(proj))] #this is required due to a bug in ArchR that was fixed in later version\n",
    "\n",
    "        proj = addPeakSet(ArchRProj = proj, peakSet = gr_peakset, force = T)\n",
    "        proj = addPeakMatrix(ArchRProj = proj)\n",
    "    \n",
    "        #Code end to add peak set\n",
    "        gc()\n",
    "        log_print(\"SUCCESSFUL: Adding Peak set\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Adding Peak set\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save prefiltered project\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Saving prefiltered project\")\n",
    "    \n",
    "        #Code start to save prefiltered project\n",
    "        \n",
    "        #write raw ArchR proj rds \n",
    "        saveRDS(proj, file = paste0(prefix,\".atac.archr.raw_project.\",genome,\".rds\"))\n",
    "    \n",
    "        #write raw .h5 matrix\n",
    "        mtx = getMatrixFromProject(ArchRProj = proj,useMatrix = \"PeakMatrix\")\n",
    "        rownames(mtx) = proj@peakSet$idx\n",
    "        write_dgCMatrix_h5(assay(mtx), \n",
    "                           cols_are = \"barcodes\", \n",
    "                           h5_target = paste0(prefix,\".atac.archr.raw_matrix.\",genome,\".h5\"), \n",
    "                           ref_name = prefix)\n",
    "        \n",
    "        #create metadata dataframe\n",
    "        mtx = chromVAR::addGCBias(mtx, genome = getBSgenome(genome)) #get gc content\n",
    "        idxPass <- which(proj$TSSEnrichment >= min_tss & proj$nFrags >= min_frags) #IDs of cells that will pass filter\n",
    "        meta = proj@cellColData\n",
    "        meta$PassQC = FALSE\n",
    "        meta[idxPass, \"PassQC\"] = TRUE\n",
    "        #meta$GCBias = mtx@rowRanges$bias\n",
    "        meta$barcodes = rownames(meta)\n",
    "        \n",
    "        #write metadata tsv \n",
    "        write.table(x = meta, file =  paste0(prefix,\".atac.archr.barcode_metadata.\",genome,\".tsv\"), sep = \"\\t\", quote = F, row.names = F)\n",
    "        \n",
    "        rm(meta)\n",
    "        rm(mtx)\n",
    "    \n",
    "        #Code end to save prefiltered project\n",
    "        gc()\n",
    "        log_print(\"SUCCESSFUL: Saving prefiltered project\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Saving prefiltered project\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filtering Archr Project\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Filter ArchR Project\")\n",
    "    \n",
    "        #Code start to filter ArchR Project\n",
    "        \n",
    "        proj <- proj[idxPass,]\n",
    "    \n",
    "        #Code end to filter ArchR Project\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Filter ArchR Project\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Filter ArchR Project\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Post-filtered TSS Enrichment vs Unique Fragments QC plot\n",
    "tryCatch({\n",
    "        log_print(\"# Post-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "\n",
    "        # Code start to create pre-filtered TSS Enrichment vs Unique Fragments QC plot    \n",
    "        df <- as.data.frame(getCellColData(proj, select = c(\"nFrags\", \"TSSEnrichment\")))\n",
    "                                \n",
    "        #df$density = get_density(df$nFrags, df$TSSEnrichment, n = 100)\n",
    "                            \n",
    "        #obj = ggplot(data=df) + \n",
    "            #geom_point(aes(x = nFrags, y = TSSEnrichment, color = density)) +  \n",
    "            #scale_x_continuous(trans = \"log10\", breaks = scales::trans_breaks(\"log10\", function(x) 10^x), labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "            #expand_limits(x = xlabel_min, y = 0) + \n",
    "            #geom_hline(yintercept = min_tss, lty = \"dashed\") + \n",
    "            #geom_vline(xintercept = min_frags, lty = \"dashed\") +\n",
    "            #scale_colour_gradientn(colors = paletteContinuous(set = \"sambaNight\")) +\n",
    "            #xlab(label =\"Unique Fragments\") + \n",
    "            #ylab(label = \"TSS Enrichment\") + \n",
    "            #annotation_logticks(sides = \"b\")\n",
    "                                                                              \n",
    "        obj = ggplot(data=df, aes(x = nFrags, y = TSSEnrichment)) + \n",
    "            geom_pointdensity(method = \"default\") +\n",
    "            scale_colour_gradientn(colors = paletteContinuous(set = \"sambaNight\")) +\n",
    "            scale_x_continuous(trans = \"log10\", breaks = scales::trans_breaks(\"log10\", function(x) 10^x), labels = scales::trans_format(\"log10\", scales::math_format(10^.x))) +\n",
    "            expand_limits(x = xlabel_min, y = 0) + \n",
    "            geom_hline(yintercept = min_tss, lty = \"dashed\") + \n",
    "            geom_vline(xintercept = min_frags, lty = \"dashed\") +\n",
    "            xlab(label =\"Unique Fragments\") + \n",
    "            ylab(label = \"TSS Enrichment\") + \n",
    "            annotation_logticks(sides = \"b\") +\n",
    "            labs(fill=\"density\") +\n",
    "            theme(axis.title=element_text(size=14), axis.text=element_text(size=10), legend.title=element_text(size=14), legend.text=element_text(size=8.5))\n",
    "        \n",
    "        tg <- textGrob('Post-filtered TSS Enrichment vs Unique Fragments QC plot', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \"Total Barcodes: \", nCells(proj), \"\\n\", \n",
    "                        \"Median Frags: \", median(proj$nFrags), \"\\n\", \n",
    "                        \"Median TSS Enrichment: \", median(proj$TSSEnrichment), \"\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "        plot_list = list(obj)\n",
    "        obj = create_plot(plot_list, tg, sg,heights=unit(c(1.5,1.5,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "        printPNG(name = \"postfiltered_tss_by_uniq_frags\", plotObject = obj, papermill = papermill, wf = 11, hf = 12)\n",
    "            \n",
    "        # Code end to create Post-filtered TSS Enrichment vs Unique Fragments QC plot \n",
    "\n",
    "        log_print(\"SUCCESSFUL: Post-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Post-filtered TSS Enrichment vs Unique Fragments QC plot\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Post-filtered Fragment Size Distribution plot\n",
    "tryCatch({\n",
    "        log_print(\"# Post-filtered Fragment Size Distribution plot\")\n",
    "\n",
    "        # Code start to create pre-filtered Fragment Size Distribution    \n",
    "        obj = plotFragmentSizes(ArchRProj = proj) +\n",
    "        theme_gray()\n",
    "            \n",
    "        tg <- textGrob('Post-filtered Fragment Size Distribution', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \"Total Barcodes: \", nCells(proj), \"\\n\", \n",
    "                        \"Total nFrags: \", paste(format(round(sum(proj$nFrags) / 1e6, 1), trim = TRUE), \"M\"), \"\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "        plot_list = list(obj)\n",
    "        obj = create_plot(plot_list, tg, sg,heights=unit(c(1.5,1.5,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "        printPNG(name = \"postfiltered_frag_size_dist\", plotObject = obj, papermill = papermill, wf = 11, hf = 12)\n",
    "            \n",
    "        # Code end to create pre-filtered Fragment Size Distribution \n",
    "\n",
    "        log_print(\"SUCCESSFUL: Post-filtered Fragment Size Distribution plot\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Post-filtered Fragment Size Distribution plot\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "doublet-scores"
    ]
   },
   "outputs": [],
   "source": [
    "#Calculate doublet scores\n",
    "if(find_doublets){\n",
    "    \n",
    "    tryCatch({\n",
    "\n",
    "        # Code start to calculate doublet scores\n",
    "\n",
    "            log_print(\"# Calculate doublet scores\")\n",
    "            proj <- addDoubletScores(\n",
    "                input = proj,\n",
    "                k = doublet_k,\n",
    "                knnMethod = doublet_knn_method,\n",
    "                LSIMethod = lsi_method\n",
    "            )\n",
    "            log_print(\"SUCCESSFUL: Calculate doublet scores\")\n",
    "\n",
    "        # Code end to calculate doublet scores\n",
    "\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Calculate doublet scores\")\n",
    "        log_print(cond)\n",
    "    })\n",
    "    \n",
    "} else{\n",
    "    log_print(\"#Skipping doublet detection step\")\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Uncomment next line to filter doublets\n",
    "\n",
    "#proj <- filterDoublets(ArchRProj = proj) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "iterative-lsi"
    ]
   },
   "outputs": [],
   "source": [
    "#Add Iterative LSI\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Add Iterative LSI\")\n",
    "    \n",
    "        #Code start to add Iterative LSI\n",
    "    \n",
    "        proj <- addIterativeLSI(ArchRProj = proj, useMatrix = iter_LSI_matrix, name = \"IterativeLSI\")\n",
    "    \n",
    "        #Code end to add Iterative LSI\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Add Iterative LSI\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Add Iterative LSI\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "add-clusters"
    ]
   },
   "outputs": [],
   "source": [
    "#Add clusters\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Add clusters\")\n",
    "    \n",
    "        #Code start to add clusters\n",
    "    \n",
    "        proj <- addClusters(input = proj, reducedDims = \"IterativeLSI\") #from the ArchR docs, IterativeLSI is the only supported option for reducedDims \n",
    "    \n",
    "        #Code end to add clusters\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Add clusters\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Add clusters\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "scrolled": true,
    "tags": [
     "add-umap"
    ]
   },
   "outputs": [],
   "source": [
    "#Add UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Add UMAP\")\n",
    "    \n",
    "        #Code start to add clusters\n",
    "    \n",
    "        proj <- addUMAP(ArchRProj = proj, reducedDims = \"IterativeLSI\") \n",
    "    \n",
    "        #Code end to add clusters\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Add UMAP\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Add UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Plot UMAP\")\n",
    "    \n",
    "        #Code start to plot UMAP - Clusters\n",
    "    \n",
    "        obj <- plotEmbedding(ArchRProj = proj, colorBy = \"cellColData\", name = \"Clusters\", embedding = \"UMAP\") + \n",
    "                geom_point(size = 0.2)+\n",
    "                theme_gray()  \n",
    "        tg <- textGrob('UMAP', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                        \" Colored by: Clusters using Louvain\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "        plot_list = list(obj)\n",
    "        \n",
    "        obj = create_plot(plot_list, tg, sg, heights=unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "    \n",
    "        printPNG('umap_clusters', obj, papermill, wf = 11, hf = 12)\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Plot UMAP\")\n",
    "    \n",
    "        #Code end to plot UMAP\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Plot UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Plot UMAP\")\n",
    "    \n",
    "    \n",
    "        #Code start to plot UMAP - nFrags\n",
    "        proj$lognFrags = log10(proj$nFrags)\n",
    "        obj <- plotEmbedding(ArchRProj = proj, colorBy = \"cellColData\", name = \"lognFrags\", embedding = \"UMAP\", \n",
    "                             pal = ArchRPalettes$purpleOrange) + \n",
    "                geom_point(size = 0.2)+\n",
    "                theme_gray()\n",
    "        tg <- textGrob('UMAP', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                        \" Colored by: log10(nFrags)\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "        plot_list = list(obj)\n",
    "        \n",
    "        obj = create_plot(plot_list, tg, sg, heights=unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "    \n",
    "        printPNG('umap_num_frags', obj, papermill, wf = 11, hf = 12)\n",
    "    \n",
    "        #Code end to plot UMAP\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Plot UMAP\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Plot UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Plot UMAP\")\n",
    "    \n",
    "        #Code start to plot UMAP - TSS Enrichment\n",
    "        obj <- plotEmbedding(ArchRProj = proj, colorBy = \"cellColData\", name = \"TSSEnrichment\", embedding = \"UMAP\",\n",
    "                            pal = ArchRPalettes$purpleOrange) + \n",
    "                geom_point(size = 0.2)+\n",
    "                theme_gray()  \n",
    "        tg <- textGrob('UMAP', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                        \" Colored by: TSSEnrichment\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "        plot_list = list(obj)\n",
    "        \n",
    "        obj = create_plot(plot_list, tg, sg, heights=unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "    \n",
    "        printPNG('umap_tss_score', obj, papermill, wf = 11, hf = 12)\n",
    "    \n",
    "        #Code end to plot UMAP\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Plot UMAP\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Plot UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Plot UMAP\")\n",
    "    \n",
    "        #Code start to plot UMAP - FRIP\n",
    "        obj <- plotEmbedding(ArchRProj = proj, colorBy = \"cellColData\", name = \"FRIP\", embedding = \"UMAP\",\n",
    "                            pal = ArchRPalettes$purpleOrange) + \n",
    "                geom_point(size = 0.2)+\n",
    "                theme_gray()  \n",
    "        tg <- textGrob('UMAP', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                        \" Colored by: FRIP\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "        plot_list = list(obj)\n",
    "        \n",
    "        obj = create_plot(plot_list, tg, sg, heights=unit(c(1,1,8), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "    \n",
    "        printPNG('umap_frip', obj, papermill, wf = 11, hf = 12)\n",
    "    \n",
    "        #Code end to plot UMAP\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Plot UMAP\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Plot UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Plot UMAP\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Plot UMAP\")\n",
    "    \n",
    "        #Code start to plot UMAP - Doublet\n",
    "        if(find_doublets){\n",
    "            doub_flag = FALSE\n",
    "            proj$DoubletScore[is.na(proj$DoubletScore)] <- 0\n",
    "            if (sum(proj$DoubletScore) > 0){\n",
    "                doub_flag = TRUE\n",
    "            }\n",
    "            obj <- plotEmbedding(ArchRProj = proj, colorBy = \"cellColData\", name = \"DoubletScore\", embedding = \"UMAP\") + \n",
    "                    geom_point(size = 0.1)+\n",
    "                    theme_gray()  \n",
    "            tg <- textGrob('UMAP', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "            sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                            \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                            \" Colored by: DoubletScore\\n\", \n",
    "                            \" Doublets detected: \", doub_flag ,\"\\n\", \n",
    "                            \"------------------------------------ \\n\"), \n",
    "                            gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "            plot_list = list(obj)\n",
    "\n",
    "            obj = create_plot(plot_list, tg, sg, heights=unit(c(1.5,1.5,7), rep(\"in\",3)), width = 10)\n",
    "            grid.draw(obj)\n",
    "\n",
    "            printPNG('umap_doublets', obj, papermill, wf = 11, hf = 12)\n",
    "\n",
    "            #Code end to plot UMAP\n",
    "            grid.newpage()\n",
    "        }\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Plot UMAP\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Plot UMAP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "extract-markers"
    ]
   },
   "outputs": [],
   "source": [
    "#Extract marker genes\n",
    "## single core worked, multicore breaks \n",
    "\n",
    "markersGS = tryCatch({\n",
    "        log_print(\"# Extract marker genes\")\n",
    "    \n",
    "        #Code start to extract marker genes\n",
    "    \n",
    "        markersGS <- getMarkerFeatures(\n",
    "            ArchRProj = proj, \n",
    "            useMatrix = \"GeneScoreMatrix\", \n",
    "            groupBy = \"Clusters\",\n",
    "            bias = c(\"TSSEnrichment\", \"log10(nFrags)\"),\n",
    "            testMethod = marker_features_test\n",
    "        )\n",
    "\n",
    "        #Code end to extract marker genes\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Extract marker genes\")\n",
    "        return(markersGS)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Extract marker genes\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "heatmap"
    ]
   },
   "outputs": [],
   "source": [
    "#Plot heatmap of upregulated genes in clusters\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Heatmap\")\n",
    "    \n",
    "        #Code start to plot heatmap\n",
    "    \n",
    "        hm <- plotMarkerHeatmap(markersGS, transpose = heatmap_transpose, nLabel = heatmap_label_n, \n",
    "                                 cutOff = heatmap_cutoff, plotLog2FC = TRUE) \n",
    "        obj = grid.grabExpr(draw(hm))\n",
    "        tg <- textGrob('Heatmap', gp = gpar(fontsize = 25, fontface = 'bold', col = 'red'))\n",
    "        sg <- textGrob(paste0(\"Library: \", prefix, \"\\n\", \n",
    "                        \" Number of Barcodes: \", nCells(proj),\"\\n\", \n",
    "                        \"------------------------------------ \\n\"), \n",
    "                        gp = gpar(fontsize = 18, fontface = 'bold'))\n",
    "\n",
    "        plot_list = list(obj)\n",
    "        \n",
    "        obj = create_plot(plot_list, tg, sg, heights=unit(c(1,1,10), rep(\"in\",3)), width = 10)\n",
    "        grid.draw(obj)\n",
    "    \n",
    "        printPNG('heatmap', obj, papermill, wf = 11, hf = 12)\n",
    "\n",
    "        #Code end to plot heatmap\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Heatmap\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Heatmap\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": [
     "plots-zip"
    ]
   },
   "outputs": [],
   "source": [
    "#Create final output files\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Final output files\")\n",
    "    \n",
    "        #Code start to create final output files\n",
    "    \n",
    "        files2zip <- dir(plot_filename, full.names = TRUE)\n",
    "        zip(zipfile = paste0(plot_filename,'.zip'), files = files2zip)\n",
    "\n",
    "        saveRDS(proj, file = paste0(prefix,\".atac.archr.filtered_project.\",genome,\".rds\"))\n",
    "    \n",
    "        mtx = getMatrixFromProject(ArchRProj = proj,useMatrix = \"PeakMatrix\")\n",
    "        rownames(mtx) = proj@peakSet$idx\n",
    "        write_dgCMatrix_h5(assay(mtx), \n",
    "                           cols_are = \"barcodes\", \n",
    "                           h5_target = paste0(prefix,\".atac.archr.filtered_matrix.\",genome,\".h5\"), \n",
    "                           ref_name = prefix)\n",
    "        rm(mtx)\n",
    "\n",
    "        #Code end to create final output files\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Final output files\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Final output files\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "log_close()"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
