{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "703817fc",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "#DORCs parameters\n",
    "rnaRDS = \"\" # .h5 file \n",
    "atacFragFile = \"\" #.tsv fragment file\n",
    "peakFile = \"\" #.bed peakset file\n",
    "genome = \"\" #genome\n",
    "nCores = 4\n",
    "savePlotsToDir = TRUE\n",
    "prefix = \"prefix\"\n",
    "\n",
    "#RNA QC parameters\n",
    "minFeature_RNA = 200 #Seurat QC for number of min features\n",
    "maxFeature_RNA = 2500 #Seurat QC for number of max features\n",
    "percentMT_RNA = 5 #Seurat QC for max % of mt \n",
    "minCells_RNA = 3 #Seurat QC for min number of cells\n",
    "\n",
    "#ATAC QC parameter\n",
    "fripCutOff = 0.3 #QC threshold for fRIP score\n",
    "chunkSize = 50000 #chunk size (number of pairs) to parallelize centering ATAC counts \n",
    "\n",
    "#Background correlation parameters\n",
    "numNearestNeighbor = 100 #Number of nearest neighbors\n",
    "numBackgroundPairs = 1e+05 #Number of background gene-peak pairs to generate\n",
    "\n",
    "#DORC genes parameters\n",
    "windowPadSize = 50000 #Regulatory region around TSS. Default is +/- 50Kb\n",
    "dorcGeneCutOff = 10 #No. sig peaks needed to be called a DORC\n",
    "corrPVal = 0.05 #pval cutoff for correlation statistical test\n",
    "topNGene = 20 #Label top N genes in j-Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c35618c2",
   "metadata": {
    "tags": [
     "r-libraries"
    ]
   },
   "outputs": [],
   "source": [
    "packages = c(\"dplyr\",\"Seurat\",\"patchwork\",\"GenomicRanges\",\"ggplot2\",\"ggrepel\",\"reshape2\",\"ggrastr\",\"BuenColors\",\"foreach\",\"iterators\",\"parallel\",\"Biostrings\",\"logr\")\n",
    "\n",
    "if(genome == \"hg38\"){\n",
    "    BiocManager::install(\"BSgenome.Hsapiens.UCSC.hg38\", update=F, ask=F)\n",
    "    packages = c(packages, \"BSgenome.Hsapiens.UCSC.hg38\")\n",
    "} else if(genome == \"mm10\"){\n",
    "    BiocManager::install(\"BSgenome.Mmusculus.UCSC.mm10\", update=F, ask=F)\n",
    "    packages = c(packages, \"BSgenome.Mmusculus.UCSC.mm10\")\n",
    "}\n",
    "\n",
    "new.packages <- packages[!(packages %in% installed.packages()[,\"Package\"])]\n",
    "if(length(new.packages)) BiocManager::install(new.packages, update=F, ask=F)\n",
    "\n",
    "suppressMessages(lapply(packages, library, character.only = TRUE))\n",
    "\n",
    "options(\"logr.notes\" = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977794b9",
   "metadata": {
    "tags": [
     "source-functions"
    ]
   },
   "outputs": [],
   "source": [
    "#Source functions \n",
    "\n",
    "#download from gh and source?\n",
    "source(\"/home/R/DORCS_helper_functions_optimized.R\")\n",
    "load(\"/home/R/TSSRanges.RData\")\n",
    "\n",
    "logfile <- file.path(paste0(prefix,\".dorcs.logfile.\",genome,\".txt\"))\n",
    "lf <- log_open(logfile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dd9fa6a",
   "metadata": {
    "tags": [
     "preprocess-rna"
    ]
   },
   "outputs": [],
   "source": [
    "#Create and preprocess RNA count matrix; using Seurat functions\n",
    "\n",
    "rnaCounts = tryCatch({\n",
    "        log_print(\"# Create RNA matrix\")\n",
    "    \n",
    "        #Code start to create RNA matrix\n",
    "    \n",
    "        rnaCounts = Read10X_h5(rnaCountMatrix)\n",
    "        rnaCounts = CreateSeuratObject(counts = rnaCounts, project = \"shareseq\", min.cells = minCells_RNA, min.features = minFeature_RNA)\n",
    "        rnaCounts[[\"percent.mt\"]] = PercentageFeatureSet(rnaCounts, pattern = \"^MT-\")\n",
    "        RNAVlnPlot = VlnPlot(rnaCounts, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3)\n",
    "        print(RNAVlnPlot)\n",
    "    \n",
    "        #rnaCounts = readRDS(rnaRDS)\n",
    "\n",
    "        #QC filtering and normalization\n",
    "        rnaCounts = subset(rnaCounts, subset = nFeature_RNA > minFeature_RNA & nFeature_RNA < maxFeature_RNA & percent.mt < percentMT_RNA)\n",
    "        rnaCounts = NormalizeData(rnaCounts)\n",
    "    \n",
    "        #Code end to create RNA matrix\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Create RNA matrix\")\n",
    "        return(rnaCounts)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Create RNA matrix\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90fe6130",
   "metadata": {
    "tags": [
     "create-atac-matrix"
    ]
   },
   "outputs": [],
   "source": [
    "#Create scATACseq matrix\n",
    "\n",
    "peaksSE = tryCatch({\n",
    "        log_print(\"# Create scATACseq matrix\")\n",
    "    \n",
    "        #Code start to create scATACseq matrix\n",
    "    \n",
    "        peaksSE = getCountsFromFrags(fragFile=atacFragFile, peakFile=peakFile)\n",
    "        \n",
    "        #Code end to create scATACseq matrix\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Create scATACseq matrix\")\n",
    "        return(peaksSE)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Create scATACseq matrix\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "542b0ec6",
   "metadata": {
    "tags": [
     "find-common-cells"
    ]
   },
   "outputs": [],
   "source": [
    "# Find common cells in RNA and ATAC; clean up before correlation step \n",
    "\n",
    "#filter by fRiP\n",
    "SE.filt = tryCatch({\n",
    "        log_print(\"# Filter by fRiP\")\n",
    "    \n",
    "        #Code start to filter by fRiP\n",
    "    \n",
    "        SE.filt = peaksSE[,peaksSE$FRIP > fripCutOff]\n",
    "        \n",
    "        #Code end to filter by fRiP\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Filter by fRiP\")\n",
    "        return(SE.filt)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Filter by fRiP\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "#extract RNA count matrix\n",
    "rnaMat = tryCatch({\n",
    "        log_print(\"# Extract RNA count matrix\")\n",
    "    \n",
    "        #Code start to extract RNA count matrix\n",
    "    \n",
    "        rnaMat <- rnaCounts[[\"RNA\"]]@data\n",
    "        \n",
    "        #Code end to extract RNA count matrix\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Extract RNA count matrix\")\n",
    "        return(rnaMat)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Extract RNA count matrix\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "\n",
    "#clean up\n",
    "tryCatch({\n",
    "        log_print(\"# Clean up\")\n",
    "    \n",
    "        #Code start to clean up\n",
    "    \n",
    "        rm(peaksSE)\n",
    "        rm(rnaCounts)\n",
    "\n",
    "        #Cleaning - change barcode names\n",
    "        SE.filt$sample = sub(\",P1\\\\.[0-9]+\", \"\", SE.filt$sample)\n",
    "        colnames(SE.filt) = sub(\",P1\\\\.[0-9]+\", \"\", colnames(SE.filt))\n",
    "        colnames(SE.filt) = gsub(\",\", \"\\\\.\", colnames(SE.filt))\n",
    "        colnames(rnaMat) = sub(\"\\\\,P1\\\\.[0-9]+\", \"\", colnames(rnaMat))\n",
    "        colnames(rnaMat) = gsub(\",\", \"\\\\.\", colnames(rnaMat))\n",
    "        \n",
    "        #Code end to clean up\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Clean up\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Clean up\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "#Get intersect of cells in RNA amd ATAC\n",
    "cells = tryCatch({\n",
    "        log_print(\"# Get intersect of cells in RNA amd ATAC\")\n",
    "    \n",
    "        #Code start to get intersect of cells in RNA amd ATAC\n",
    "    \n",
    "        cells = intersect(colnames(SE.filt), colnames(rnaMat))\n",
    "        \n",
    "        #Code end to get intersect of cells in RNA amd ATAC\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Get intersect of cells in RNA amd ATAC\")\n",
    "        return(cells)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Get intersect of cells in RNA amd ATAC\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5844c936",
   "metadata": {
    "tags": [
     "correlation"
    ]
   },
   "outputs": [],
   "source": [
    "#Correlation\n",
    "\n",
    "cisCor = tryCatch({\n",
    "        log_print(\"# Correlation\")\n",
    "    \n",
    "        #Code start to run correlation\n",
    "    \n",
    "        set.seed(123)\n",
    "\n",
    "        #Run fast gene peak correlation\n",
    "        cisCor <- fastGenePeakcorr(\n",
    "          SE.filt[,cells],\n",
    "          rnaMat[,cells],\n",
    "          genome = genome, # This will be one of \"hg19\",\"hg38\" or \"mm10\"\n",
    "          windowPadSize = windowPadSize,\n",
    "          normalizeATACmat = TRUE,\n",
    "          nCores = nCores,\n",
    "          p.cut = NULL,\n",
    "          n_bg = numNearestNeighbor,\n",
    "          n_BgPairs = numBackgroundPairs,\n",
    "          chunkSize = chunkSize\n",
    "        )\n",
    "        \n",
    "        #Code end to run correlation\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Correlation\")\n",
    "        return(cisCor)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Correlation\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "#filter by pval\n",
    "cisCor.filt <- cisCor %>% dplyr::filter(pvalZ <= corrPVal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8125e787",
   "metadata": {
    "tags": [
     "find-dorcs"
    ]
   },
   "outputs": [],
   "source": [
    "#dorcGenes and j-Plot\n",
    "\n",
    "dorcGenes = tryCatch({\n",
    "        log_print(\"# DORC genes and J-plot\")\n",
    "    \n",
    "        #Code start to find DORC genes and create j-Plot\n",
    "    \n",
    "        dorcGenes = dorcJPlot(dorcTab = cisCor.filt,\n",
    "                       cutoff = dorcGeneCutOff, # No. sig peaks needed to be called a DORC\n",
    "                       labelTop = topNGene,\n",
    "                       returnGeneList = TRUE, # Set this to FALSE for just the plot\n",
    "                       force=2)\n",
    "    \n",
    "        #Code end to find DORC genes and create j-Plot\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: DORC genes and J-plot\")\n",
    "        return(dorcGenes)\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: DORC genes and J-plot\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d11964b7",
   "metadata": {
    "tags": [
     "plots-zip"
    ]
   },
   "outputs": [],
   "source": [
    "#Create final output files\n",
    "\n",
    "tryCatch({\n",
    "        log_print(\"# Final output files\")\n",
    "    \n",
    "        #Code start to create final output files\n",
    "    \n",
    "        if(savePlotsToDir){\n",
    "            dir.create(\"plots\")\n",
    "            savePlots = function(name, plotObject){\n",
    "                filename = paste0(prefix,\".dorcs.\",name,\".\",genome)\n",
    "                png(sprintf(\"plots/%s.png\", filename), width=480*wf, height=480*hf)\n",
    "                print(plotObject)\n",
    "                dev.off()\n",
    "            }\n",
    "            savePlots(\"rna_violin_plot\", RNAVlnPlot)\n",
    "            savePlots(\"jplot\", dorcJPlot(dorcTab = cisCor.filt,cutoff = dorcGeneCutOff, labelTop = topNGene,returnGeneList = FALSE,force=2))\n",
    "        }\n",
    "\n",
    "        write.table(dorcGenes, file=paste0(prefix,\".dorcs.\",\"dorc_genes_summary\",\".\",genome,\".csv\"), row.names = T, quote = F, sep = \",\")\n",
    "        write.table(cisCor, file=paste0(prefix,\".dorcs.\",\"all_regions_summary\",\".\",genome,\".csv\"), row.names = T, quote = F, sep = \",\")\n",
    "\n",
    "        files2zip <- dir('plots', full.names = TRUE)\n",
    "        zip(zipfile = 'plots', files = files2zip)\n",
    "\n",
    "        #Code end to create final output files\n",
    "    \n",
    "        log_print(\"SUCCESSFUL: Final output files\")\n",
    "    },\n",
    "    error = function(cond) {\n",
    "        log_print(\"ERROR: Final output files\")\n",
    "        log_print(cond)\n",
    "    }\n",
    ")\n",
    "\n",
    "log_close()"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
