{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "703817fc",
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "#Might need to change filepaths\n",
    "\n",
    "rnaRDS = \"../../dorcs_data_example/out.gene.bc.matrices.h5\"\n",
    "atacFragFile = \"../../dorcs_data_example/shareseq-project.atac.GRCh38.cleaned.filtered.bedpe\"\n",
    "peakFile = \"../../dorcs_data_example/GM_nonoverlap.bed\"\n",
    "cellKNNMatrix = \"../../dorcs_data_example/cellKNN_GM_nonoverlap.rds\"\n",
    "pwmFile = \"../R/cisBP_human_pfms_2021.rds\" #need better name, but don't know original reasoning\n",
    "\n",
    "genome = \"hg38\"\n",
    "nCores = 4\n",
    "savePlotsToDir = TRUE\n",
    "\n",
    "minFeature_RNA = 200 #Seurat QC for number of min features\n",
    "maxFeature_RNA = 2500 #Seurat QC for number of max features\n",
    "percentMT_RNA = 5 #Seurat QC for max % of mt \n",
    "minCells_RNA = 3 #Seurat QC for min number of cells\n",
    "\n",
    "dorcGeneCutOff = 10 #No. sig peaks needed to be called a DORC\n",
    "fripCutOff = 0.3 #QC threshold for fRIP score\n",
    "do = 0.05 #pval cutoff for correlation statistical test\n",
    "nDORCknn = 10 #No. of dorc kNNs used to pool peaks\n",
    "topNGene = 20 #Label top N genes in j-Plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "c35618c2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cran.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.14 (BiocManager 1.30.16), R 4.1.2 (2021-11-01)\n",
      "\n",
      "Warning message:\n",
      "“package(s) not installed when version(s) same as current; use `force = TRUE` to\n",
      "  re-install: 'BSgenome.Hsapiens.UCSC.hg38'”\n",
      "Old packages: 'Cairo', 'clipr', 'commonmark', 'desc', 'DT', 'evaluate',\n",
      "  'fitdistrplus', 'float', 'future', 'GenomicFeatures', 'gmp', 'limma', 'mgcv',\n",
      "  'misty', 'mockery', 'pracma', 'Rcpp', 'renv', 'rhdf5', 'rmarkdown',\n",
      "  'survival', 'tinytex', 'uuid', 'waldo', 'xfun', 'yaml'\n",
      "\n",
      "also installing the dependencies ‘backports’, ‘bayestestR’, ‘datawizard’, ‘insight’, ‘RUnit’, ‘broom’, ‘broom.mixed’, ‘parameters’, ‘performance’, ‘carData’, ‘kutils’, ‘broomExtra’, ‘rockchalk’, ‘r2mlm’, ‘writexl’\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  There are binary versions available but the source versions are later:\n",
      "       binary  source needs_compilation\n",
      "Cairo  1.5-14  1.5-15              TRUE\n",
      "gmp     0.6-4   0.6-5              TRUE\n",
      "Rcpp  1.0.8.2 1.0.8.3              TRUE\n",
      "uuid    1.0-3   1.0-4              TRUE\n",
      "waldo   0.3.1   0.4.0             FALSE\n",
      "\n",
      "\n",
      "The downloaded binary packages are in\n",
      "\t/var/folders/c3/cr8rqqls3gn0r7h2cq16rxg80000gp/T//Rtmp1ye5ht/downloaded_packages\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "installing the source packages ‘Cairo’, ‘gmp’, ‘Rcpp’, ‘uuid’, ‘waldo’\n",
      "\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<ol>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'iterators'</li><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'parallel'</li><li>'iterators'</li><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'Biostrings'</li><li>'XVector'</li><li>'parallel'</li><li>'iterators'</li><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'rtracklayer'</li><li>'Biostrings'</li><li>'XVector'</li><li>'parallel'</li><li>'iterators'</li><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "\t<li><style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'BSgenome.Hsapiens.UCSC.hg38'</li><li>'BSgenome'</li><li>'rtracklayer'</li><li>'Biostrings'</li><li>'XVector'</li><li>'parallel'</li><li>'iterators'</li><li>'foreach'</li><li>'motifmatchr'</li><li>'network'</li><li>'GGally'</li><li>'networkD3'</li><li>'circlize'</li><li>'ComplexHeatmap'</li><li>'grid'</li><li>'BuenColors'</li><li>'MASS'</li><li>'ggrastr'</li><li>'reshape2'</li><li>'ggrepel'</li><li>'ggplot2'</li><li>'GenomicRanges'</li><li>'GenomeInfoDb'</li><li>'IRanges'</li><li>'S4Vectors'</li><li>'BiocGenerics'</li><li>'stats4'</li><li>'patchwork'</li><li>'SeuratObject'</li><li>'Seurat'</li><li>'dplyr'</li><li>'stats'</li><li>'graphics'</li><li>'grDevices'</li><li>'utils'</li><li>'datasets'</li><li>'methods'</li><li>'base'</li></ol>\n",
       "</li>\n",
       "</ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate}\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'iterators'\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'parallel'\n",
       "\\item 'iterators'\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'Biostrings'\n",
       "\\item 'XVector'\n",
       "\\item 'parallel'\n",
       "\\item 'iterators'\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'rtracklayer'\n",
       "\\item 'Biostrings'\n",
       "\\item 'XVector'\n",
       "\\item 'parallel'\n",
       "\\item 'iterators'\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\item \\begin{enumerate*}\n",
       "\\item 'BSgenome.Hsapiens.UCSC.hg38'\n",
       "\\item 'BSgenome'\n",
       "\\item 'rtracklayer'\n",
       "\\item 'Biostrings'\n",
       "\\item 'XVector'\n",
       "\\item 'parallel'\n",
       "\\item 'iterators'\n",
       "\\item 'foreach'\n",
       "\\item 'motifmatchr'\n",
       "\\item 'network'\n",
       "\\item 'GGally'\n",
       "\\item 'networkD3'\n",
       "\\item 'circlize'\n",
       "\\item 'ComplexHeatmap'\n",
       "\\item 'grid'\n",
       "\\item 'BuenColors'\n",
       "\\item 'MASS'\n",
       "\\item 'ggrastr'\n",
       "\\item 'reshape2'\n",
       "\\item 'ggrepel'\n",
       "\\item 'ggplot2'\n",
       "\\item 'GenomicRanges'\n",
       "\\item 'GenomeInfoDb'\n",
       "\\item 'IRanges'\n",
       "\\item 'S4Vectors'\n",
       "\\item 'BiocGenerics'\n",
       "\\item 'stats4'\n",
       "\\item 'patchwork'\n",
       "\\item 'SeuratObject'\n",
       "\\item 'Seurat'\n",
       "\\item 'dplyr'\n",
       "\\item 'stats'\n",
       "\\item 'graphics'\n",
       "\\item 'grDevices'\n",
       "\\item 'utils'\n",
       "\\item 'datasets'\n",
       "\\item 'methods'\n",
       "\\item 'base'\n",
       "\\end{enumerate*}\n",
       "\n",
       "\\end{enumerate}\n"
      ],
      "text/markdown": [
       "1. 1. 'dplyr'\n",
       "2. 'stats'\n",
       "3. 'graphics'\n",
       "4. 'grDevices'\n",
       "5. 'utils'\n",
       "6. 'datasets'\n",
       "7. 'methods'\n",
       "8. 'base'\n",
       "\n",
       "\n",
       "\n",
       "2. 1. 'SeuratObject'\n",
       "2. 'Seurat'\n",
       "3. 'dplyr'\n",
       "4. 'stats'\n",
       "5. 'graphics'\n",
       "6. 'grDevices'\n",
       "7. 'utils'\n",
       "8. 'datasets'\n",
       "9. 'methods'\n",
       "10. 'base'\n",
       "\n",
       "\n",
       "\n",
       "3. 1. 'patchwork'\n",
       "2. 'SeuratObject'\n",
       "3. 'Seurat'\n",
       "4. 'dplyr'\n",
       "5. 'stats'\n",
       "6. 'graphics'\n",
       "7. 'grDevices'\n",
       "8. 'utils'\n",
       "9. 'datasets'\n",
       "10. 'methods'\n",
       "11. 'base'\n",
       "\n",
       "\n",
       "\n",
       "4. 1. 'GenomicRanges'\n",
       "2. 'GenomeInfoDb'\n",
       "3. 'IRanges'\n",
       "4. 'S4Vectors'\n",
       "5. 'BiocGenerics'\n",
       "6. 'stats4'\n",
       "7. 'patchwork'\n",
       "8. 'SeuratObject'\n",
       "9. 'Seurat'\n",
       "10. 'dplyr'\n",
       "11. 'stats'\n",
       "12. 'graphics'\n",
       "13. 'grDevices'\n",
       "14. 'utils'\n",
       "15. 'datasets'\n",
       "16. 'methods'\n",
       "17. 'base'\n",
       "\n",
       "\n",
       "\n",
       "5. 1. 'ggplot2'\n",
       "2. 'GenomicRanges'\n",
       "3. 'GenomeInfoDb'\n",
       "4. 'IRanges'\n",
       "5. 'S4Vectors'\n",
       "6. 'BiocGenerics'\n",
       "7. 'stats4'\n",
       "8. 'patchwork'\n",
       "9. 'SeuratObject'\n",
       "10. 'Seurat'\n",
       "11. 'dplyr'\n",
       "12. 'stats'\n",
       "13. 'graphics'\n",
       "14. 'grDevices'\n",
       "15. 'utils'\n",
       "16. 'datasets'\n",
       "17. 'methods'\n",
       "18. 'base'\n",
       "\n",
       "\n",
       "\n",
       "6. 1. 'ggrepel'\n",
       "2. 'ggplot2'\n",
       "3. 'GenomicRanges'\n",
       "4. 'GenomeInfoDb'\n",
       "5. 'IRanges'\n",
       "6. 'S4Vectors'\n",
       "7. 'BiocGenerics'\n",
       "8. 'stats4'\n",
       "9. 'patchwork'\n",
       "10. 'SeuratObject'\n",
       "11. 'Seurat'\n",
       "12. 'dplyr'\n",
       "13. 'stats'\n",
       "14. 'graphics'\n",
       "15. 'grDevices'\n",
       "16. 'utils'\n",
       "17. 'datasets'\n",
       "18. 'methods'\n",
       "19. 'base'\n",
       "\n",
       "\n",
       "\n",
       "7. 1. 'reshape2'\n",
       "2. 'ggrepel'\n",
       "3. 'ggplot2'\n",
       "4. 'GenomicRanges'\n",
       "5. 'GenomeInfoDb'\n",
       "6. 'IRanges'\n",
       "7. 'S4Vectors'\n",
       "8. 'BiocGenerics'\n",
       "9. 'stats4'\n",
       "10. 'patchwork'\n",
       "11. 'SeuratObject'\n",
       "12. 'Seurat'\n",
       "13. 'dplyr'\n",
       "14. 'stats'\n",
       "15. 'graphics'\n",
       "16. 'grDevices'\n",
       "17. 'utils'\n",
       "18. 'datasets'\n",
       "19. 'methods'\n",
       "20. 'base'\n",
       "\n",
       "\n",
       "\n",
       "8. 1. 'ggrastr'\n",
       "2. 'reshape2'\n",
       "3. 'ggrepel'\n",
       "4. 'ggplot2'\n",
       "5. 'GenomicRanges'\n",
       "6. 'GenomeInfoDb'\n",
       "7. 'IRanges'\n",
       "8. 'S4Vectors'\n",
       "9. 'BiocGenerics'\n",
       "10. 'stats4'\n",
       "11. 'patchwork'\n",
       "12. 'SeuratObject'\n",
       "13. 'Seurat'\n",
       "14. 'dplyr'\n",
       "15. 'stats'\n",
       "16. 'graphics'\n",
       "17. 'grDevices'\n",
       "18. 'utils'\n",
       "19. 'datasets'\n",
       "20. 'methods'\n",
       "21. 'base'\n",
       "\n",
       "\n",
       "\n",
       "9. 1. 'BuenColors'\n",
       "2. 'MASS'\n",
       "3. 'ggrastr'\n",
       "4. 'reshape2'\n",
       "5. 'ggrepel'\n",
       "6. 'ggplot2'\n",
       "7. 'GenomicRanges'\n",
       "8. 'GenomeInfoDb'\n",
       "9. 'IRanges'\n",
       "10. 'S4Vectors'\n",
       "11. 'BiocGenerics'\n",
       "12. 'stats4'\n",
       "13. 'patchwork'\n",
       "14. 'SeuratObject'\n",
       "15. 'Seurat'\n",
       "16. 'dplyr'\n",
       "17. 'stats'\n",
       "18. 'graphics'\n",
       "19. 'grDevices'\n",
       "20. 'utils'\n",
       "21. 'datasets'\n",
       "22. 'methods'\n",
       "23. 'base'\n",
       "\n",
       "\n",
       "\n",
       "10. 1. 'ComplexHeatmap'\n",
       "2. 'grid'\n",
       "3. 'BuenColors'\n",
       "4. 'MASS'\n",
       "5. 'ggrastr'\n",
       "6. 'reshape2'\n",
       "7. 'ggrepel'\n",
       "8. 'ggplot2'\n",
       "9. 'GenomicRanges'\n",
       "10. 'GenomeInfoDb'\n",
       "11. 'IRanges'\n",
       "12. 'S4Vectors'\n",
       "13. 'BiocGenerics'\n",
       "14. 'stats4'\n",
       "15. 'patchwork'\n",
       "16. 'SeuratObject'\n",
       "17. 'Seurat'\n",
       "18. 'dplyr'\n",
       "19. 'stats'\n",
       "20. 'graphics'\n",
       "21. 'grDevices'\n",
       "22. 'utils'\n",
       "23. 'datasets'\n",
       "24. 'methods'\n",
       "25. 'base'\n",
       "\n",
       "\n",
       "\n",
       "11. 1. 'circlize'\n",
       "2. 'ComplexHeatmap'\n",
       "3. 'grid'\n",
       "4. 'BuenColors'\n",
       "5. 'MASS'\n",
       "6. 'ggrastr'\n",
       "7. 'reshape2'\n",
       "8. 'ggrepel'\n",
       "9. 'ggplot2'\n",
       "10. 'GenomicRanges'\n",
       "11. 'GenomeInfoDb'\n",
       "12. 'IRanges'\n",
       "13. 'S4Vectors'\n",
       "14. 'BiocGenerics'\n",
       "15. 'stats4'\n",
       "16. 'patchwork'\n",
       "17. 'SeuratObject'\n",
       "18. 'Seurat'\n",
       "19. 'dplyr'\n",
       "20. 'stats'\n",
       "21. 'graphics'\n",
       "22. 'grDevices'\n",
       "23. 'utils'\n",
       "24. 'datasets'\n",
       "25. 'methods'\n",
       "26. 'base'\n",
       "\n",
       "\n",
       "\n",
       "12. 1. 'networkD3'\n",
       "2. 'circlize'\n",
       "3. 'ComplexHeatmap'\n",
       "4. 'grid'\n",
       "5. 'BuenColors'\n",
       "6. 'MASS'\n",
       "7. 'ggrastr'\n",
       "8. 'reshape2'\n",
       "9. 'ggrepel'\n",
       "10. 'ggplot2'\n",
       "11. 'GenomicRanges'\n",
       "12. 'GenomeInfoDb'\n",
       "13. 'IRanges'\n",
       "14. 'S4Vectors'\n",
       "15. 'BiocGenerics'\n",
       "16. 'stats4'\n",
       "17. 'patchwork'\n",
       "18. 'SeuratObject'\n",
       "19. 'Seurat'\n",
       "20. 'dplyr'\n",
       "21. 'stats'\n",
       "22. 'graphics'\n",
       "23. 'grDevices'\n",
       "24. 'utils'\n",
       "25. 'datasets'\n",
       "26. 'methods'\n",
       "27. 'base'\n",
       "\n",
       "\n",
       "\n",
       "13. 1. 'GGally'\n",
       "2. 'networkD3'\n",
       "3. 'circlize'\n",
       "4. 'ComplexHeatmap'\n",
       "5. 'grid'\n",
       "6. 'BuenColors'\n",
       "7. 'MASS'\n",
       "8. 'ggrastr'\n",
       "9. 'reshape2'\n",
       "10. 'ggrepel'\n",
       "11. 'ggplot2'\n",
       "12. 'GenomicRanges'\n",
       "13. 'GenomeInfoDb'\n",
       "14. 'IRanges'\n",
       "15. 'S4Vectors'\n",
       "16. 'BiocGenerics'\n",
       "17. 'stats4'\n",
       "18. 'patchwork'\n",
       "19. 'SeuratObject'\n",
       "20. 'Seurat'\n",
       "21. 'dplyr'\n",
       "22. 'stats'\n",
       "23. 'graphics'\n",
       "24. 'grDevices'\n",
       "25. 'utils'\n",
       "26. 'datasets'\n",
       "27. 'methods'\n",
       "28. 'base'\n",
       "\n",
       "\n",
       "\n",
       "14. 1. 'network'\n",
       "2. 'GGally'\n",
       "3. 'networkD3'\n",
       "4. 'circlize'\n",
       "5. 'ComplexHeatmap'\n",
       "6. 'grid'\n",
       "7. 'BuenColors'\n",
       "8. 'MASS'\n",
       "9. 'ggrastr'\n",
       "10. 'reshape2'\n",
       "11. 'ggrepel'\n",
       "12. 'ggplot2'\n",
       "13. 'GenomicRanges'\n",
       "14. 'GenomeInfoDb'\n",
       "15. 'IRanges'\n",
       "16. 'S4Vectors'\n",
       "17. 'BiocGenerics'\n",
       "18. 'stats4'\n",
       "19. 'patchwork'\n",
       "20. 'SeuratObject'\n",
       "21. 'Seurat'\n",
       "22. 'dplyr'\n",
       "23. 'stats'\n",
       "24. 'graphics'\n",
       "25. 'grDevices'\n",
       "26. 'utils'\n",
       "27. 'datasets'\n",
       "28. 'methods'\n",
       "29. 'base'\n",
       "\n",
       "\n",
       "\n",
       "15. 1. 'motifmatchr'\n",
       "2. 'network'\n",
       "3. 'GGally'\n",
       "4. 'networkD3'\n",
       "5. 'circlize'\n",
       "6. 'ComplexHeatmap'\n",
       "7. 'grid'\n",
       "8. 'BuenColors'\n",
       "9. 'MASS'\n",
       "10. 'ggrastr'\n",
       "11. 'reshape2'\n",
       "12. 'ggrepel'\n",
       "13. 'ggplot2'\n",
       "14. 'GenomicRanges'\n",
       "15. 'GenomeInfoDb'\n",
       "16. 'IRanges'\n",
       "17. 'S4Vectors'\n",
       "18. 'BiocGenerics'\n",
       "19. 'stats4'\n",
       "20. 'patchwork'\n",
       "21. 'SeuratObject'\n",
       "22. 'Seurat'\n",
       "23. 'dplyr'\n",
       "24. 'stats'\n",
       "25. 'graphics'\n",
       "26. 'grDevices'\n",
       "27. 'utils'\n",
       "28. 'datasets'\n",
       "29. 'methods'\n",
       "30. 'base'\n",
       "\n",
       "\n",
       "\n",
       "16. 1. 'foreach'\n",
       "2. 'motifmatchr'\n",
       "3. 'network'\n",
       "4. 'GGally'\n",
       "5. 'networkD3'\n",
       "6. 'circlize'\n",
       "7. 'ComplexHeatmap'\n",
       "8. 'grid'\n",
       "9. 'BuenColors'\n",
       "10. 'MASS'\n",
       "11. 'ggrastr'\n",
       "12. 'reshape2'\n",
       "13. 'ggrepel'\n",
       "14. 'ggplot2'\n",
       "15. 'GenomicRanges'\n",
       "16. 'GenomeInfoDb'\n",
       "17. 'IRanges'\n",
       "18. 'S4Vectors'\n",
       "19. 'BiocGenerics'\n",
       "20. 'stats4'\n",
       "21. 'patchwork'\n",
       "22. 'SeuratObject'\n",
       "23. 'Seurat'\n",
       "24. 'dplyr'\n",
       "25. 'stats'\n",
       "26. 'graphics'\n",
       "27. 'grDevices'\n",
       "28. 'utils'\n",
       "29. 'datasets'\n",
       "30. 'methods'\n",
       "31. 'base'\n",
       "\n",
       "\n",
       "\n",
       "17. 1. 'iterators'\n",
       "2. 'foreach'\n",
       "3. 'motifmatchr'\n",
       "4. 'network'\n",
       "5. 'GGally'\n",
       "6. 'networkD3'\n",
       "7. 'circlize'\n",
       "8. 'ComplexHeatmap'\n",
       "9. 'grid'\n",
       "10. 'BuenColors'\n",
       "11. 'MASS'\n",
       "12. 'ggrastr'\n",
       "13. 'reshape2'\n",
       "14. 'ggrepel'\n",
       "15. 'ggplot2'\n",
       "16. 'GenomicRanges'\n",
       "17. 'GenomeInfoDb'\n",
       "18. 'IRanges'\n",
       "19. 'S4Vectors'\n",
       "20. 'BiocGenerics'\n",
       "21. 'stats4'\n",
       "22. 'patchwork'\n",
       "23. 'SeuratObject'\n",
       "24. 'Seurat'\n",
       "25. 'dplyr'\n",
       "26. 'stats'\n",
       "27. 'graphics'\n",
       "28. 'grDevices'\n",
       "29. 'utils'\n",
       "30. 'datasets'\n",
       "31. 'methods'\n",
       "32. 'base'\n",
       "\n",
       "\n",
       "\n",
       "18. 1. 'parallel'\n",
       "2. 'iterators'\n",
       "3. 'foreach'\n",
       "4. 'motifmatchr'\n",
       "5. 'network'\n",
       "6. 'GGally'\n",
       "7. 'networkD3'\n",
       "8. 'circlize'\n",
       "9. 'ComplexHeatmap'\n",
       "10. 'grid'\n",
       "11. 'BuenColors'\n",
       "12. 'MASS'\n",
       "13. 'ggrastr'\n",
       "14. 'reshape2'\n",
       "15. 'ggrepel'\n",
       "16. 'ggplot2'\n",
       "17. 'GenomicRanges'\n",
       "18. 'GenomeInfoDb'\n",
       "19. 'IRanges'\n",
       "20. 'S4Vectors'\n",
       "21. 'BiocGenerics'\n",
       "22. 'stats4'\n",
       "23. 'patchwork'\n",
       "24. 'SeuratObject'\n",
       "25. 'Seurat'\n",
       "26. 'dplyr'\n",
       "27. 'stats'\n",
       "28. 'graphics'\n",
       "29. 'grDevices'\n",
       "30. 'utils'\n",
       "31. 'datasets'\n",
       "32. 'methods'\n",
       "33. 'base'\n",
       "\n",
       "\n",
       "\n",
       "19. 1. 'Biostrings'\n",
       "2. 'XVector'\n",
       "3. 'parallel'\n",
       "4. 'iterators'\n",
       "5. 'foreach'\n",
       "6. 'motifmatchr'\n",
       "7. 'network'\n",
       "8. 'GGally'\n",
       "9. 'networkD3'\n",
       "10. 'circlize'\n",
       "11. 'ComplexHeatmap'\n",
       "12. 'grid'\n",
       "13. 'BuenColors'\n",
       "14. 'MASS'\n",
       "15. 'ggrastr'\n",
       "16. 'reshape2'\n",
       "17. 'ggrepel'\n",
       "18. 'ggplot2'\n",
       "19. 'GenomicRanges'\n",
       "20. 'GenomeInfoDb'\n",
       "21. 'IRanges'\n",
       "22. 'S4Vectors'\n",
       "23. 'BiocGenerics'\n",
       "24. 'stats4'\n",
       "25. 'patchwork'\n",
       "26. 'SeuratObject'\n",
       "27. 'Seurat'\n",
       "28. 'dplyr'\n",
       "29. 'stats'\n",
       "30. 'graphics'\n",
       "31. 'grDevices'\n",
       "32. 'utils'\n",
       "33. 'datasets'\n",
       "34. 'methods'\n",
       "35. 'base'\n",
       "\n",
       "\n",
       "\n",
       "20. 1. 'rtracklayer'\n",
       "2. 'Biostrings'\n",
       "3. 'XVector'\n",
       "4. 'parallel'\n",
       "5. 'iterators'\n",
       "6. 'foreach'\n",
       "7. 'motifmatchr'\n",
       "8. 'network'\n",
       "9. 'GGally'\n",
       "10. 'networkD3'\n",
       "11. 'circlize'\n",
       "12. 'ComplexHeatmap'\n",
       "13. 'grid'\n",
       "14. 'BuenColors'\n",
       "15. 'MASS'\n",
       "16. 'ggrastr'\n",
       "17. 'reshape2'\n",
       "18. 'ggrepel'\n",
       "19. 'ggplot2'\n",
       "20. 'GenomicRanges'\n",
       "21. 'GenomeInfoDb'\n",
       "22. 'IRanges'\n",
       "23. 'S4Vectors'\n",
       "24. 'BiocGenerics'\n",
       "25. 'stats4'\n",
       "26. 'patchwork'\n",
       "27. 'SeuratObject'\n",
       "28. 'Seurat'\n",
       "29. 'dplyr'\n",
       "30. 'stats'\n",
       "31. 'graphics'\n",
       "32. 'grDevices'\n",
       "33. 'utils'\n",
       "34. 'datasets'\n",
       "35. 'methods'\n",
       "36. 'base'\n",
       "\n",
       "\n",
       "\n",
       "21. 1. 'BSgenome.Hsapiens.UCSC.hg38'\n",
       "2. 'BSgenome'\n",
       "3. 'rtracklayer'\n",
       "4. 'Biostrings'\n",
       "5. 'XVector'\n",
       "6. 'parallel'\n",
       "7. 'iterators'\n",
       "8. 'foreach'\n",
       "9. 'motifmatchr'\n",
       "10. 'network'\n",
       "11. 'GGally'\n",
       "12. 'networkD3'\n",
       "13. 'circlize'\n",
       "14. 'ComplexHeatmap'\n",
       "15. 'grid'\n",
       "16. 'BuenColors'\n",
       "17. 'MASS'\n",
       "18. 'ggrastr'\n",
       "19. 'reshape2'\n",
       "20. 'ggrepel'\n",
       "21. 'ggplot2'\n",
       "22. 'GenomicRanges'\n",
       "23. 'GenomeInfoDb'\n",
       "24. 'IRanges'\n",
       "25. 'S4Vectors'\n",
       "26. 'BiocGenerics'\n",
       "27. 'stats4'\n",
       "28. 'patchwork'\n",
       "29. 'SeuratObject'\n",
       "30. 'Seurat'\n",
       "31. 'dplyr'\n",
       "32. 'stats'\n",
       "33. 'graphics'\n",
       "34. 'grDevices'\n",
       "35. 'utils'\n",
       "36. 'datasets'\n",
       "37. 'methods'\n",
       "38. 'base'\n",
       "\n",
       "\n",
       "\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[[1]]\n",
       "[1] \"dplyr\"     \"stats\"     \"graphics\"  \"grDevices\" \"utils\"     \"datasets\" \n",
       "[7] \"methods\"   \"base\"     \n",
       "\n",
       "[[2]]\n",
       " [1] \"SeuratObject\" \"Seurat\"       \"dplyr\"        \"stats\"        \"graphics\"    \n",
       " [6] \"grDevices\"    \"utils\"        \"datasets\"     \"methods\"      \"base\"        \n",
       "\n",
       "[[3]]\n",
       " [1] \"patchwork\"    \"SeuratObject\" \"Seurat\"       \"dplyr\"        \"stats\"       \n",
       " [6] \"graphics\"     \"grDevices\"    \"utils\"        \"datasets\"     \"methods\"     \n",
       "[11] \"base\"        \n",
       "\n",
       "[[4]]\n",
       " [1] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [5] \"BiocGenerics\"  \"stats4\"        \"patchwork\"     \"SeuratObject\" \n",
       " [9] \"Seurat\"        \"dplyr\"         \"stats\"         \"graphics\"     \n",
       "[13] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[17] \"base\"         \n",
       "\n",
       "[[5]]\n",
       " [1] \"ggplot2\"       \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"      \n",
       " [5] \"S4Vectors\"     \"BiocGenerics\"  \"stats4\"        \"patchwork\"    \n",
       " [9] \"SeuratObject\"  \"Seurat\"        \"dplyr\"         \"stats\"        \n",
       "[13] \"graphics\"      \"grDevices\"     \"utils\"         \"datasets\"     \n",
       "[17] \"methods\"       \"base\"         \n",
       "\n",
       "[[6]]\n",
       " [1] \"ggrepel\"       \"ggplot2\"       \"GenomicRanges\" \"GenomeInfoDb\" \n",
       " [5] \"IRanges\"       \"S4Vectors\"     \"BiocGenerics\"  \"stats4\"       \n",
       " [9] \"patchwork\"     \"SeuratObject\"  \"Seurat\"        \"dplyr\"        \n",
       "[13] \"stats\"         \"graphics\"      \"grDevices\"     \"utils\"        \n",
       "[17] \"datasets\"      \"methods\"       \"base\"         \n",
       "\n",
       "[[7]]\n",
       " [1] \"reshape2\"      \"ggrepel\"       \"ggplot2\"       \"GenomicRanges\"\n",
       " [5] \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"     \"BiocGenerics\" \n",
       " [9] \"stats4\"        \"patchwork\"     \"SeuratObject\"  \"Seurat\"       \n",
       "[13] \"dplyr\"         \"stats\"         \"graphics\"      \"grDevices\"    \n",
       "[17] \"utils\"         \"datasets\"      \"methods\"       \"base\"         \n",
       "\n",
       "[[8]]\n",
       " [1] \"ggrastr\"       \"reshape2\"      \"ggrepel\"       \"ggplot2\"      \n",
       " [5] \"GenomicRanges\" \"GenomeInfoDb\"  \"IRanges\"       \"S4Vectors\"    \n",
       " [9] \"BiocGenerics\"  \"stats4\"        \"patchwork\"     \"SeuratObject\" \n",
       "[13] \"Seurat\"        \"dplyr\"         \"stats\"         \"graphics\"     \n",
       "[17] \"grDevices\"     \"utils\"         \"datasets\"      \"methods\"      \n",
       "[21] \"base\"         \n",
       "\n",
       "[[9]]\n",
       " [1] \"BuenColors\"    \"MASS\"          \"ggrastr\"       \"reshape2\"     \n",
       " [5] \"ggrepel\"       \"ggplot2\"       \"GenomicRanges\" \"GenomeInfoDb\" \n",
       " [9] \"IRanges\"       \"S4Vectors\"     \"BiocGenerics\"  \"stats4\"       \n",
       "[13] \"patchwork\"     \"SeuratObject\"  \"Seurat\"        \"dplyr\"        \n",
       "[17] \"stats\"         \"graphics\"      \"grDevices\"     \"utils\"        \n",
       "[21] \"datasets\"      \"methods\"       \"base\"         \n",
       "\n",
       "[[10]]\n",
       " [1] \"ComplexHeatmap\" \"grid\"           \"BuenColors\"     \"MASS\"          \n",
       " [5] \"ggrastr\"        \"reshape2\"       \"ggrepel\"        \"ggplot2\"       \n",
       " [9] \"GenomicRanges\"  \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"     \n",
       "[13] \"BiocGenerics\"   \"stats4\"         \"patchwork\"      \"SeuratObject\"  \n",
       "[17] \"Seurat\"         \"dplyr\"          \"stats\"          \"graphics\"      \n",
       "[21] \"grDevices\"      \"utils\"          \"datasets\"       \"methods\"       \n",
       "[25] \"base\"          \n",
       "\n",
       "[[11]]\n",
       " [1] \"circlize\"       \"ComplexHeatmap\" \"grid\"           \"BuenColors\"    \n",
       " [5] \"MASS\"           \"ggrastr\"        \"reshape2\"       \"ggrepel\"       \n",
       " [9] \"ggplot2\"        \"GenomicRanges\"  \"GenomeInfoDb\"   \"IRanges\"       \n",
       "[13] \"S4Vectors\"      \"BiocGenerics\"   \"stats4\"         \"patchwork\"     \n",
       "[17] \"SeuratObject\"   \"Seurat\"         \"dplyr\"          \"stats\"         \n",
       "[21] \"graphics\"       \"grDevices\"      \"utils\"          \"datasets\"      \n",
       "[25] \"methods\"        \"base\"          \n",
       "\n",
       "[[12]]\n",
       " [1] \"networkD3\"      \"circlize\"       \"ComplexHeatmap\" \"grid\"          \n",
       " [5] \"BuenColors\"     \"MASS\"           \"ggrastr\"        \"reshape2\"      \n",
       " [9] \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\"  \"GenomeInfoDb\"  \n",
       "[13] \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"   \"stats4\"        \n",
       "[17] \"patchwork\"      \"SeuratObject\"   \"Seurat\"         \"dplyr\"         \n",
       "[21] \"stats\"          \"graphics\"       \"grDevices\"      \"utils\"         \n",
       "[25] \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[13]]\n",
       " [1] \"GGally\"         \"networkD3\"      \"circlize\"       \"ComplexHeatmap\"\n",
       " [5] \"grid\"           \"BuenColors\"     \"MASS\"           \"ggrastr\"       \n",
       " [9] \"reshape2\"       \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\" \n",
       "[13] \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"  \n",
       "[17] \"stats4\"         \"patchwork\"      \"SeuratObject\"   \"Seurat\"        \n",
       "[21] \"dplyr\"          \"stats\"          \"graphics\"       \"grDevices\"     \n",
       "[25] \"utils\"          \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[14]]\n",
       " [1] \"network\"        \"GGally\"         \"networkD3\"      \"circlize\"      \n",
       " [5] \"ComplexHeatmap\" \"grid\"           \"BuenColors\"     \"MASS\"          \n",
       " [9] \"ggrastr\"        \"reshape2\"       \"ggrepel\"        \"ggplot2\"       \n",
       "[13] \"GenomicRanges\"  \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"     \n",
       "[17] \"BiocGenerics\"   \"stats4\"         \"patchwork\"      \"SeuratObject\"  \n",
       "[21] \"Seurat\"         \"dplyr\"          \"stats\"          \"graphics\"      \n",
       "[25] \"grDevices\"      \"utils\"          \"datasets\"       \"methods\"       \n",
       "[29] \"base\"          \n",
       "\n",
       "[[15]]\n",
       " [1] \"motifmatchr\"    \"network\"        \"GGally\"         \"networkD3\"     \n",
       " [5] \"circlize\"       \"ComplexHeatmap\" \"grid\"           \"BuenColors\"    \n",
       " [9] \"MASS\"           \"ggrastr\"        \"reshape2\"       \"ggrepel\"       \n",
       "[13] \"ggplot2\"        \"GenomicRanges\"  \"GenomeInfoDb\"   \"IRanges\"       \n",
       "[17] \"S4Vectors\"      \"BiocGenerics\"   \"stats4\"         \"patchwork\"     \n",
       "[21] \"SeuratObject\"   \"Seurat\"         \"dplyr\"          \"stats\"         \n",
       "[25] \"graphics\"       \"grDevices\"      \"utils\"          \"datasets\"      \n",
       "[29] \"methods\"        \"base\"          \n",
       "\n",
       "[[16]]\n",
       " [1] \"foreach\"        \"motifmatchr\"    \"network\"        \"GGally\"        \n",
       " [5] \"networkD3\"      \"circlize\"       \"ComplexHeatmap\" \"grid\"          \n",
       " [9] \"BuenColors\"     \"MASS\"           \"ggrastr\"        \"reshape2\"      \n",
       "[13] \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\"  \"GenomeInfoDb\"  \n",
       "[17] \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"   \"stats4\"        \n",
       "[21] \"patchwork\"      \"SeuratObject\"   \"Seurat\"         \"dplyr\"         \n",
       "[25] \"stats\"          \"graphics\"       \"grDevices\"      \"utils\"         \n",
       "[29] \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[17]]\n",
       " [1] \"iterators\"      \"foreach\"        \"motifmatchr\"    \"network\"       \n",
       " [5] \"GGally\"         \"networkD3\"      \"circlize\"       \"ComplexHeatmap\"\n",
       " [9] \"grid\"           \"BuenColors\"     \"MASS\"           \"ggrastr\"       \n",
       "[13] \"reshape2\"       \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\" \n",
       "[17] \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"  \n",
       "[21] \"stats4\"         \"patchwork\"      \"SeuratObject\"   \"Seurat\"        \n",
       "[25] \"dplyr\"          \"stats\"          \"graphics\"       \"grDevices\"     \n",
       "[29] \"utils\"          \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[18]]\n",
       " [1] \"parallel\"       \"iterators\"      \"foreach\"        \"motifmatchr\"   \n",
       " [5] \"network\"        \"GGally\"         \"networkD3\"      \"circlize\"      \n",
       " [9] \"ComplexHeatmap\" \"grid\"           \"BuenColors\"     \"MASS\"          \n",
       "[13] \"ggrastr\"        \"reshape2\"       \"ggrepel\"        \"ggplot2\"       \n",
       "[17] \"GenomicRanges\"  \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"     \n",
       "[21] \"BiocGenerics\"   \"stats4\"         \"patchwork\"      \"SeuratObject\"  \n",
       "[25] \"Seurat\"         \"dplyr\"          \"stats\"          \"graphics\"      \n",
       "[29] \"grDevices\"      \"utils\"          \"datasets\"       \"methods\"       \n",
       "[33] \"base\"          \n",
       "\n",
       "[[19]]\n",
       " [1] \"Biostrings\"     \"XVector\"        \"parallel\"       \"iterators\"     \n",
       " [5] \"foreach\"        \"motifmatchr\"    \"network\"        \"GGally\"        \n",
       " [9] \"networkD3\"      \"circlize\"       \"ComplexHeatmap\" \"grid\"          \n",
       "[13] \"BuenColors\"     \"MASS\"           \"ggrastr\"        \"reshape2\"      \n",
       "[17] \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\"  \"GenomeInfoDb\"  \n",
       "[21] \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"   \"stats4\"        \n",
       "[25] \"patchwork\"      \"SeuratObject\"   \"Seurat\"         \"dplyr\"         \n",
       "[29] \"stats\"          \"graphics\"       \"grDevices\"      \"utils\"         \n",
       "[33] \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[20]]\n",
       " [1] \"rtracklayer\"    \"Biostrings\"     \"XVector\"        \"parallel\"      \n",
       " [5] \"iterators\"      \"foreach\"        \"motifmatchr\"    \"network\"       \n",
       " [9] \"GGally\"         \"networkD3\"      \"circlize\"       \"ComplexHeatmap\"\n",
       "[13] \"grid\"           \"BuenColors\"     \"MASS\"           \"ggrastr\"       \n",
       "[17] \"reshape2\"       \"ggrepel\"        \"ggplot2\"        \"GenomicRanges\" \n",
       "[21] \"GenomeInfoDb\"   \"IRanges\"        \"S4Vectors\"      \"BiocGenerics\"  \n",
       "[25] \"stats4\"         \"patchwork\"      \"SeuratObject\"   \"Seurat\"        \n",
       "[29] \"dplyr\"          \"stats\"          \"graphics\"       \"grDevices\"     \n",
       "[33] \"utils\"          \"datasets\"       \"methods\"        \"base\"          \n",
       "\n",
       "[[21]]\n",
       " [1] \"BSgenome.Hsapiens.UCSC.hg38\" \"BSgenome\"                   \n",
       " [3] \"rtracklayer\"                 \"Biostrings\"                 \n",
       " [5] \"XVector\"                     \"parallel\"                   \n",
       " [7] \"iterators\"                   \"foreach\"                    \n",
       " [9] \"motifmatchr\"                 \"network\"                    \n",
       "[11] \"GGally\"                      \"networkD3\"                  \n",
       "[13] \"circlize\"                    \"ComplexHeatmap\"             \n",
       "[15] \"grid\"                        \"BuenColors\"                 \n",
       "[17] \"MASS\"                        \"ggrastr\"                    \n",
       "[19] \"reshape2\"                    \"ggrepel\"                    \n",
       "[21] \"ggplot2\"                     \"GenomicRanges\"              \n",
       "[23] \"GenomeInfoDb\"                \"IRanges\"                    \n",
       "[25] \"S4Vectors\"                   \"BiocGenerics\"               \n",
       "[27] \"stats4\"                      \"patchwork\"                  \n",
       "[29] \"SeuratObject\"                \"Seurat\"                     \n",
       "[31] \"dplyr\"                       \"stats\"                      \n",
       "[33] \"graphics\"                    \"grDevices\"                  \n",
       "[35] \"utils\"                       \"datasets\"                   \n",
       "[37] \"methods\"                     \"base\"                       \n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "packages = c(\"dplyr\",\"Seurat\",\"patchwork\",\"GenomicRanges\",\"ggplot2\",\"ggrepel\",\"reshape2\",\"ggrastr\",\"BuenColors\",\"ComplexHeatmap\", \"circlize\",\"networkD3\",\"GGally\",\"network\",\"motifmatchr\",\"foreach\",\"iterators\",\"parallel\",\"Biostrings\",\"rtracklayer\")\n",
    "\n",
    "if(genome == \"hg38\"){\n",
    "    BiocManager::install(\"BSgenome.Hsapiens.UCSC.hg38\", update=T, ask=F)\n",
    "    packages = c(packages, \"BSgenome.Hsapiens.UCSC.hg38\")\n",
    "} else if(genome == \"mm10\"){\n",
    "    BiocManager::install(\"BSgenome.Mmusculus.UCSC.mm10\", update=T, ask=F)\n",
    "    packages = c(packages, \"BSgenome.Mmusculus.UCSC.mm10\")\n",
    "}\n",
    "\n",
    "new.packages <- packages[!(packages %in% installed.packages()[,\"Package\"])]\n",
    "if(length(new.packages)) BiocManager::install(new.packages, update=T, ask=F)\n",
    "\n",
    "suppressMessages(lapply(packages, library, character.only = TRUE))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "977794b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Might need to change filepaths\n",
    "#download from gh and source?\n",
    "#cellkNN = readRDS(cellKNNMatrix)\n",
    "source(\"/R/DORCS_helper_functions.R\")\n",
    "source(\"/R/FigR_functions.R\")\n",
    "load(\"/R/TSSRanges.RData\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dd9fa6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create and preprocess RNA count matrix; using Seurat functions\n",
    "\n",
    "#rnaCounts = Read10X_h5(rnaCountMatrix)\n",
    "#rnaCounts = CreateSeuratObject(counts = rnaCounts, project = \"shareseq\", min.cells = minCells_RNA, min.features = minFeature_RNA)\n",
    "#rnaCounts[[\"percent.mt\"]] = PercentageFeatureSet(rnaCounts, pattern = \"^MT-\")\n",
    "#RNAVlnPlot = VlnPlot(rnaCounts, features = c(\"nFeature_RNA\", \"nCount_RNA\", \"percent.mt\"), ncol = 3)\n",
    "#print(RNAVlnPlot)\n",
    "\n",
    "rnaCounts = readRDS(rnaRDS)\n",
    "#QC filtering and normalization\n",
    "\n",
    "rnaCounts = subset(rnaCounts, subset = nFeature_RNA > minFeature_RNA & nFeature_RNA < maxFeature_RNA & percent.mt < percentMT_RNA)\n",
    "rnaCounts = NormalizeData(rnaCounts)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "90fe6130",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create and preprocess scATACseq\n",
    "\n",
    "myPeaks = read.table(peakFile, sep=\"\\t\", header=F)\n",
    "myFrags = read.table(atacFragFile, sep=\"\\t\", header=F)u\n",
    "\n",
    "#myFrags = fread(\"data/merged.fragments.100k.tsv\", sep=\"\\t\", header=F)\n",
    "\n",
    "peakRanges = makeGRangesFromDataFrame(myPeaks,seqnames.field = \"V1\",start.field = \"V2\",end.field = \"V3\",starts.in.df.are.0based = TRUE)\n",
    "fragRanges = makeGRangesFromDataFrame(myFrags, seqnames.field = \"V1\", start.field = \"V2\", end.field = \"V3\", keep.extra.columns = TRUE, starts.in.df.are.0based = TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "542b0ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get counts from fragfile\n",
    "peaksSE = getCountsFromFrags(fragRanges=fragRanges, peaks=peakRanges )\n",
    "SE.filt = peaksSE[,peaksSE$FRIP > fripCutOff]\n",
    "\n",
    "#extract RNA count matrix\n",
    "rnaMat <- rnaCounts[[\"RNA\"]]@data\n",
    "\n",
    "#Cleaning - change barcode names\n",
    "SE.filt$sample = sub(\",P1\\\\.[0-9]+\", \"\", SE.filt$sample)\n",
    "colnames(SE.filt) = sub(\",P1\\\\.[0-9]+\", \"\", colnames(SE.filt))\n",
    "colnames(SE.filt) = gsub(\",\", \"\\\\.\", colnames(SE.filt))\n",
    "colnames(rnaMat) = sub(\"\\\\,P1\\\\.[0-9]+\", \"\", colnames(rnaMat))\n",
    "colnames(rnaMat) = gsub(\",\", \"\\\\.\", colnames(rnaMat))\n",
    "\n",
    "#Get intersect of cells in RNA amd ATAC\n",
    "cells = intersect(colnames(SE.filt), colnames(rnaMat))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5844c936",
   "metadata": {},
   "outputs": [],
   "source": [
    "set.seed(123)\n",
    "\n",
    "#Run fast gene peak correlation\n",
    "cisCor <- fastGenePeakcorr(\n",
    "  SE.filt[,cells],\n",
    "  rnaMat[,cells],\n",
    "  genome = genome, # This will be one of \"hg19\",\"hg38\" or \"mm10\"\n",
    "  windowPadSize = 50000,\n",
    "  normalizeATACmat = TRUE,\n",
    "  nCores = nCores,\n",
    "  p.cut = NULL\n",
    ")\n",
    "\n",
    "cisCor.filt <- cisCor %>% filter(pvalZ <= corrPVal)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8125e787",
   "metadata": {},
   "outputs": [],
   "source": [
    "#dorcGenes and j-Plot\n",
    "dorcGenes = dorcJPlot(dorcTab = cisCor.filt,\n",
    "                       cutoff = dorcGeneCutOff, # No. sig peaks needed to be called a DORC\n",
    "                       labelTop = topNGene,\n",
    "                       returnGeneList = TRUE, # Set this to FALSE for just the plot\n",
    "                       force=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d11964b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "if(savePlotsToDir){\n",
    "    dir.create(\"/plots\")\n",
    "    savePlots = function(filename, plotObject){\n",
    "        pdf(paste0(\"/plots/\",filename,\".pdf\"))\n",
    "        print(plotObject)\n",
    "        dev.off()\n",
    "        \n",
    "        png(paste0(\"/plots/\",filename,\".png\"))\n",
    "        print(plotObject)\n",
    "        dev.off()\n",
    "    }\n",
    "    #savePlots(\"RNAViolinPlot\", RNAVlnPlot)\n",
    "    savePlots(\"JPlot\", dorcJPlot(dorcTab = cisCor.filt,cutoff = dorcGeneCutOff, labelTop = topNGene,returnGeneList = FALSE,force=2))\n",
    "    #savePlots(\"figRPlot\", figRPlot)\n",
    "    #savePlots(\"figRHeatmap\", figRHeatmap)\n",
    "\n",
    "}\n",
    "\n",
    "#saveRDS(figR.d, file = \"figR_object.rds\")\n",
    "\n",
    "files2zip <- dir('/plots', full.names = TRUE)\n",
    "zip(zipfile = '/plots', files = files2zip)\n",
    "\n",
    "BiocManager::install(c(\"BSgenome.Hsapiens.UCSC.hg38\",\"GenomicDistributions\"), update=T, ask=F)\n",
    "myFrags = fread(\"data/merged.fragments.tsv\", header =F, sep = \"\\t\", showProgress=T)"
   ]
  }
 ],
 "metadata": {
  "celltoolbar": "Tags",
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
