#!/usr/bin/Rscript 

### This script takes the scRNA-seq data generated by SHARE-seq and
### perform cell annotation using Seurat label transfer approach.

## Import helper functions
source("/usr/local/bin/cell_annotation_helper_functions.R")

suppressMessages(library(Seurat))
suppressMessages(library(anndata))
suppressMessages(library(reticulate))
suppressMessages(library(Matrix))
suppressMessages(library(future))
suppressMessages(library(logr))
suppressMessages(library(dplyr))
suppressMessages(library(grid))
suppressMessages(library(gridExtra))
suppressMessages(library(ggplot2))
suppressMessages(library(patchwork))
suppressMessages(library(cowplot))
suppressMessages(library(EnsDb.Mmusculus.v79))
suppressMessages(library(EnsDb.Hsapiens.v86))
suppressMessages(library(optparse))

options("logr.notes" = FALSE)
options(future.globals.maxSize=10e9)
set.seed(1234)

option_list = list(
    make_option(c("--prefix"), type="character", default="prefix", 
                help="Sample or project name", 
                metavar="character"),
    make_option(c("--reference_data_name"), type="character", default="reference.h5ad", 
                help="Reference data. Must be a .h5ad", 
                metavar="character"),
    make_option(c("--reference_label"), type="character", default="cell_type", 
                help="Reference data. Must be a .h5ad", 
                metavar="character"),
    make_option(c("--query_data"), type="character", default="query.rds", 
                help="Query data.", 
                metavar="character"),
    make_option(c("--genome"), type="character", default=NULL, 
                help="Reference genome. Should be either hg38 or mm10", 
                metavar="Reference genome"),
    make_option(c("--gene_id_to_symbol"), type="character", default="TRUE", 
                help="Set true if the reference data uses gene id as feature name. 
                This is usually true for data downloaded from cellxgene server", 
                metavar="character")
)

opt_parser = OptionParser(option_list=option_list)
opt = parse_args(opt_parser)

# Get parameter
prefix = opt$prefix
reference_data_name = opt$reference_data_name
reference_label = opt$reference_label
query_data = opt$query_data
genome = opt$genome
gene_id_to_symbol = as.logical(opt$gene_id_to_symbol)

# Function to save plots
plot_filename = glue::glue("{prefix}.cell.annotation.plots.{genome}")
dir.create(plot_filename, showWarnings=F)

printPNG <- function(name, plot, papermill, width = 22, height = 11){
    filename = glue::glue("{plot_filename}/{prefix}.cell.annotation.{name}.{genome}.png")
    
    if(papermill){
        ggsave(plot = plot, filename = filename, width = width, height = height)
    }
}

# Create log file
logfile <- file.path(glue::glue("{prefix}.cell.annotation.logfile.{genome}.txt"))
lf <- log_open(logfile)


# Loading H5AD file
tryCatch(
    {
        log_print("# Loading reference data...")
        adata <- read_h5ad(reference_data_name)
        
        counts <- t(as.matrix(adata$raw$X))
        colnames(counts) <- adata$obs_names
        rownames(counts) <- adata$var_names
        
        metadata <- as.data.frame(adata$obs)

        obj.ref <- CreateSeuratObject(counts = counts, assay = "RNA")
        obj.ref <- AddMetaData(obj.ref, metadata)
        
        rm(counts)
        gc()
        
    },
    error = function(cond){
        log_print("ERROR: loading reference data")
        log_print(cond)
    
    }


)

# Read query data
tryCatch(
    {
        log_print("# Reading query data...")
        obj.query <- readRDS(query_data)
        log_print("SUCCESSFUL: Reading query data")
    
    },
    error = function(cond) {
        log_print("ERROR: Reading query data")
        log_print(cond)
    }
)


# Convert gene ID to symbol for reference data
if(gene_id_to_symbol){
    tryCatch(
    {
        log_print("# Converting gene id to symbol for reference data")
        
        if(genome == "hg38" | genome == "hg37"){
            gene.id <- ensembldb::select(EnsDb.Hsapiens.v86, 
                                         keys= rownames(obj.ref), 
                                         keytype = "GENEID", 
                                         columns = c("SYMBOL","GENEID"))
    
        } else if(genome == "mm10" | genome == "mm9"){
            gene.id <- ensembldb::select(EnsDb.Mmusculus.v79, 
                                         keys = rownames(obj.ref), 
                                         keytype = "GENEID", 
                                         columns = c("SYMBOL","GENEID"))
        }
        
        # remove genes with empty symbol
        gene.id <- subset(gene.id, gene.id$SYMBOL != "")

        # make gene symbol unique
        gene.id$Unique_SYMBOL <- make.unique(gene.id$SYMBOL, "")

        counts <- obj.ref@assays$RNA@counts
        colnames(counts) <- colnames(obj.ref)
        rownames(counts) <- rownames(obj.ref)
        
        counts <- counts[gene.id$GENEID, ]
        rownames(counts) <- gene.id$Unique_SYMBOL

        obj.ref <- CreateSeuratObject(counts = counts, 
                                      meta.data = obj.ref@meta.data)
        
        log_print("SUCCESSFUL: Converting gene id to symbol for reference data")

    },
    error = function(cond) {
        log_print("ERROR: Converting gene id to symbol for reference data")
        log_print(cond)
    }
)
}

# Subset reference data
tryCatch(
    {
        log_print("# Subseting reference and query data with common genes")
        
        gene.common <- intersect(rownames(obj.ref), rownames(obj.query))
        
        counts <- obj.ref@assays$RNA@counts[gene.common, ]
        obj.ref <- CreateSeuratObject(counts = counts, 
                                      assay = "RNA",
                                      meta.data = obj.ref@meta.data)
        
        counts <- obj.query@assays$RNA@counts[gene.common, ]
        obj.query <- CreateSeuratObject(counts = counts, 
                                        assay = "RNA",
                                        meta.data = obj.query@meta.data)
        
        
        obj.ref <- obj.ref %>%
            NormalizeData(verbose = FALSE) %>%
            FindVariableFeatures() %>%
            ScaleData() %>%
            RunPCA(verbose=FALSE) %>%
            RunUMAP(verbose=FALSE, dims=1:30)
        
        obj.query <- obj.query %>%
            NormalizeData(verbose = FALSE) %>%
            FindVariableFeatures() %>%
            ScaleData() %>%
            RunPCA(verbose=FALSE) %>%
            RunUMAP(verbose=FALSE, dims=1:30)
        
        
        log_print(glue::glue("# Found {length(gene.common)} common genes between reference and query data"))
        log_print("SUCCESSFUL: Subseting reference and query data with common genes")
    },
     error = function(cond) {
        log_print("ERROR: Subseting reference and query data with common genes")
        log_print(cond)
    }
    
)

# Predict labels for query dataset
tryCatch(
    {
        log_print("# Predicting labels for query data")
        
        transfer.anchors <- FindTransferAnchors(
            reference = obj.ref,
            query = obj.query,
            reduction = "cca",
            verbose = TRUE
        )
        
        predictions <- TransferData(anchorset = transfer.anchors, 
                                    refdata = obj.ref[[reference_label]][, 1],
                                    weight.reduction = obj.query[["pca"]],
                                    dims = 1:30,
                                    verbose = TRUE)
        
        obj.query <- AddMetaData(obj.query, metadata = predictions)

        write.csv(predictions, 
                  file = glue::glue("{prefix}.cell.annotation.prediction.{genome}.csv"),
                  quote = FALSE)
        
        log_print("SUCCESSFUL: Predicting labels for query data")
    },
     error = function(cond) {
        log_print("ERROR: Predicting labels for query data")
        log_print(cond)
    }
    
)

## Plotting
tryCatch(
    {
        log_print("# Plotting predicted labels")
        
        p <- DimPlot(obj.query, group.by = "predicted.id", label = TRUE, 
                      label.size = 5, repel = TRUE, reduction = "umap")
        
        printPNG(name = "predicted.labels", plot = p, papermill = papermill, 
                 width = 6, height = 6)
        
        log_print("SUCCESSFUL: Plotting predicted labels")
        

    },
    error = function(cond) {
        log_print("ERROR: Plotting predicted labels")
        log_print(cond)
    }

)